enterprise/backend/src/metabase_enterprise/caching/strategies.clj

```diff
@@ -0,0 +1,204 @@
+                  :limit    1
+                  :from     [[{:union-all qs} :unused_alias]]
+                  :order-by :ordering}
+          config (t2/select-one :model/CacheConfig :id q)]
+      (if config
+        (merge {:type       (:strategy config)
+                :updated_at (:updated_at config)
+                :payload    (:payload config)}
+               (:config config))
+        (when-let [ttl (granular-ttl card dashboard-id)]
+          {:type     :duration
+           :duration ttl
+           :unit     "hours"})))))
+
+;;; Strategy execution
+
+(defmulti fetch-cache-stmt-ee*
+  "Generate prepared statement for a db cache backend for a given strategy"
+  (fn [strategy _hash _conn] (:type strategy)))
+
+(defmethod fetch-cache-stmt-ee* :ttl [strategy query-hash conn]
+  (backend.db/fetch-cache-stmt-ttl strategy query-hash conn))
+
+(defmethod fetch-cache-stmt-ee* :duration [strategy query-hash conn]
+  (assert (:duration strategy))
+  (assert (:unit strategy))
+  (let [duration  (t/duration (:duration strategy) (keyword (:unit strategy)))
+        timestamp (t/minus (t/offset-date-time) duration)]
+    (backend.db/prepare-statement conn query-hash timestamp)))
+
+;; NOTE: something changes `:updated_at` to `:updated-at`
+(defmethod fetch-cache-stmt-ee* :schedule [{:keys [updated-at] :as strategy} query-hash conn]
+  (if-not updated-at
+    (log/debugf "Caching strategy %s was never run yet" (pr-str strategy))
+    (backend.db/prepare-statement conn query-hash updated-at)))
+
+(defmethod fetch-cache-stmt-ee* :query [{:keys [updated-at] :as strategy} query-hash conn]
+  (if-not updated-at
+    (log/debugf "Caching strategy %s was never run yet" (pr-str strategy))
+    (backend.db/prepare-statement conn query-hash updated-at)))
+
+(defmethod fetch-cache-stmt-ee* :nocache [_ _ _]
+  nil)
+
+(defenterprise fetch-cache-stmt
+  "Returns prepared statement to query for db cache backend."
+  :feature :cache-granular-controls
+  [strategy query-hash conn]
+  (fetch-cache-stmt-ee* strategy query-hash conn))
+
+;;; Jobs
+
+(defn- refresh-schedule-configs []
+  (let [now (t/offset-date-time)]
+    (doseq [item (t2/select :model/CacheConfig :strategy :schedule {:where [:or
+                                                                            [:= :next_run_at nil]
+                                                                            [:<= :next_run_at [:now]]]})
+            :let [cron     (-> (cron/cron-schedule (-> item :config :schedule))
+                               cron/finalize)
+                  ;; needed by the tests, or the cron will use it's own current date
+                  _        (.setStartTime ^MutableTrigger cron (t/java-date))
+                  next-run (-> (.getFireTimeAfter ^MutableTrigger cron (t/java-date (t/offset-date-time)))
+                               (t/offset-date-time (t/zone-offset)))]]
+      (t2/update! :model/CacheConfig {:id (:id item)} {:updated_at  now
```

Comment:
Would it not be better to use the database's own `:now` function? It seems dangerous to query by db time but write app time.

Reply:
Woah this is something to think about actually. Let me process it a little bit. Maybe select should happen using our time as well?..

Reply:
In the end decision was to split `updated_at` and `last_expired_at` - first one as a general marker of changes, second as a "logical time", set from Metabase rather than from database.